rust-symcrypt/src/gcm.rs

```diff
@@ -4,97 +4,119 @@ use crate::block_ciphers::*;
 use crate::errors::SymCryptError;
 use core::ffi::c_void;
 use std::mem;
+use std::pin::Pin;
 use std::ptr;
 use std::vec;
 use symcrypt_sys;
 
 pub struct GcmState {
-    expanded_key: symcrypt_sys::SYMCRYPT_GCM_EXPANDED_KEY,
+    inner: Pin<Box<GcmStateInner>>,
+}
+
+struct GcmStateInner {
     state: symcrypt_sys::SYMCRYPT_GCM_STATE,
+    expanded_key: symcrypt_sys::SYMCRYPT_GCM_EXPANDED_KEY,
 }
 
-// Have to Box<> since memory address for Self is moved around when returning from GcmState::new()
-// My guess is that it has to do with Result<> wrapping the Self and SymCryptError.
-// The only accepted Cipher for GCM is AesBlock.
+/// SymCrypt expects the address for its structs to stay static through the structs lifetime.
```

Comment:
Possibly worth noting in this comment that the reason for this is that to guarantee that the structs aren't `memcpy`'d as doing so can lead to use-after-free, inconsistent states, etc.

Reply:
Will make note